### PR TITLE
Prevent Default Source/Stock to be deleted

### PR DESCRIPTION
**Pull request by bartekszymanski**
### Fixed Issues (if relevant)

1. magento-engcom/msi#120: Prevent Default Source/Stock to be deleted

### Manual testing scenarios

1. Login to Magento admin
2. Go to Manage Stock
3. Try to delete default stock (id:1) from grid or detail page

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+     * @throws CouldNotDeleteException
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function beforeDeleteById(StockRepositoryInterface $subject, int $stockId)
+    {
+        if ($stockId === $this->defaultStockProvider->getId()) {
+            throw new CouldNotDeleteException(__('Default Stock could not be deleted.'));
+        }
+
+        return null;
```
Comment (naydav) [inline app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php]: Pls remove this return
We should return nothing (void)

File: app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+     * @param int $stockId
+     *
+     * @return null
+     * @throws CouldNotDeleteException
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function beforeDeleteById(StockRepositoryInterface $subject, int $stockId)
+    {
+        if ($stockId === $this->defaultStockProvider->getId()) {
+            throw new CouldNotDeleteException(__('Default Stock could not be deleted.'));
```
Comment (naydav) [inline app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php:44]: Need to cover this check with web-api test
Example of negative scenarios you could find in MSI web-api tests

File: app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        $this->defaultStockProvider = $defaultStockProvider;
+    }
+
+    /**
+     * Prevent default source to be deleted
+     *
+     * @param StockRepositoryInterface $subject
+     * @param int $stockId
+     *
+     * @return null
```
Comment (naydav) [inline app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php]: return void

File: app/code/Magento/InventoryCatalog/etc/di.xml [hunk 4]
```diff
@@ -8,4 +8,7 @@
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
     <preference for="Magento\InventoryCatalog\Api\DefaultStockProviderInterface" type="Magento\InventoryCatalog\Model\DefaultStockProvider"/>
     <preference for="Magento\InventoryCatalog\Api\DefaultSourceProviderInterface" type="Magento\InventoryCatalog\Model\DefaultSourceProvider"/>
+    <type name="Magento\Inventory\Model\StockRepository">
```
Comment (naydav) [inline app/code/Magento/InventoryCatalog/etc/di.xml]: We need to add this plugin not on implementation `Magento\Inventory\Model\StockRepository` but on interface `\Magento\InventoryApi\Api\StockRepositoryInterface`

If somebody will replace `StockRepository` on own implementation our check will be work

File: app/code/Magento/InventoryCatalog/etc/di.xml [hunk 5]
```diff
@@ -8,4 +8,7 @@
...
+        <plugin name="prevent-default-stock-deletion" type="Magento\InventoryCatalog\Plugin\Model\StockRepositoryPlugin" sortOrder="1"/>
```
Comment (naydav) [inline app/code/Magento/InventoryCatalog/etc/di.xml]: As I see in Magento generally used underscore for keys (but not dashe)

File: app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php (hunk 3 above)
Comment (larsroettig) [inline app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php]: why null return void is enough

File: app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+namespace Magento\InventoryCatalog\Plugin\Model;
+
+use Magento\Framework\Exception\CouldNotDeleteException;
+use Magento\InventoryApi\Api\StockRepositoryInterface;
+use Magento\InventoryCatalog\Api\DefaultStockProviderInterface;
+
+/**
+ * Plugin Stock Repository
```
Comment (larsroettig) [inline app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php]: Describe better or remove

File: app/code/Magento/InventoryCatalog/etc/di.xml [hunk 7]
```diff
@@ -8,4 +8,7 @@
...
+    </type>
```
Comment (naydav) [inline app/code/Magento/InventoryCatalog/etc/di.xml]: Also, I think we could expand Delete button object.
We do not need show delete button in admin panel if it is default Stock or Source

File: app/code/Magento/InventoryCatalog/etc/di.xml (hunk 5 above)
Comment (larsroettig) [inline app/code/Magento/InventoryCatalog/etc/di.xml]: pls use **prevent_default_stock_deletion** as **name**

**Review by larsroettig — CHANGES_REQUESTED** [review]

File: app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php (hunk 3 above)
Comment (bartekszymanski) [reply app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php]: it is from http://devdocs.magento.com/guides/v2.2/extension-dev-guide/plugins.html "If the method does not change the argument for the observed method, it should return null" but maybe my thinking was wrong in this case.

File: app/code/Magento/InventoryCatalog/etc/di.xml (hunk 5 above)
Comment (bartekszymanski) [reply app/code/Magento/InventoryCatalog/etc/di.xml]: got it, again like in magento2 original code but of course I will fix that

File: app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php (hunk 2 above)
Comment (bartekszymanski) [reply app/code/Magento/InventoryCatalog/Plugin/Model/StockRepositoryPlugin.php:44]: done

**Review by larsroettig — APPROVED** [review]

**Pull request merged**